local volume provisioner: improve helm chart

cofyc opened the pull request:
- Get rid of `engine` thing, it's better to use differnt values files for different environments.
- Rename `jobsrole` and `local-storage-provisioner-job-binding` to `local-storage-provisioner`, and only create them when necessary.
- Able to create StorageClass object and configure it.
- Add docs for all chart configurations.
  - https://github.com/cofyc/external-storage/blob/refactor_heml_chart/local-volume/helm/README.md#configurations
- Add some examples.
  - https://github.com/cofyc/external-storage/blob/refactor_heml_chart/local-volume/helm/README.md#examples

cofyc commented:
/assign @msau42 
/area local-volume

msau42 commented on local-volume/helm/README.md:
```diff
@@ -44,21 +44,46 @@ The generated specs can be further customized as needed (usually not necessary),
+| common.useJobForCleaning               | Is set to true, provisioner will use jobs-based block cleaning.                                       | bool     | `false`                                                    |
+| common.configMapName                   | Provisioner ConfigMap name.                                                                           | str      | `local-provisioner-config`                                 |
+| classes.[n].name                       | StorageClass name.                                                                                    | str      | `-`                                                        |
+| classes.[n].hostDir                    | Path on the host where local volume of this storage class are mounted.                                | str      | `-`                                                        |
```
local volumes of this storage class are mounted under

msau42 commented on local-volume/helm/README.md:
```diff
@@ -44,21 +44,46 @@ The generated specs can be further customized as needed (usually not necessary),
+| classes.[n].storageClass.reclaimPolicy | Specify reclaimPolicy of storage class, available: Delete/Retain.                                     | str      | `Delete`                                                   |
+| daemonset.name                         | Provisioner DaemonSet name.                                                                           | str      | `local-volume-provisioner`                                 |
+| daemonset.image                        | Provisioner image.                                                                                    | str      | `quay.io/external_storage/local-volume-provisioner:v2.1.0` |
+| daemonset.imagePullPolicy              | Provisioner DaemonSet image pull policy.                                                              | str      | `Always`                                                   |
```
I'm thinking it may be better to just use the default imagePullPolicy if not set.

msau42 commented on local-volume/helm/examples/gce-cleanbyjobs.yaml:
```diff
@@ -0,0 +1,15 @@
+classes:
+- name: local-scsi
+  hostDir: "/mnt/disks/by-uuid/google-local-ssds-scsi-fs"
+  blockCleanerCommand:
```
On gce/gke, we actually separate block and preformatted volumes into different directories.  So this block cleaning config doesn't do anything.  It may be better to put this in a baremetal example.

cofyc replied:
Should I remove all `blockCleanerCommand` sections from other gce/gke example yamls? I never run kubernetes on gce/gke, could you help test these examples if possible?

msau42 replied:
Yeah I would leave it out for now and add it to the baremetal example instead.  I have an issue open on me to integrate and test the block feature on gce/gke

cofyc replied:
Done in https://github.com/kubernetes-incubator/external-storage/pull/789/commits/427e9675173c9d7c734e8d2a7ca620ec077ce96d

msau42 commented on local-volume/helm/examples/baremetal-without-rbac.yaml:
```diff
@@ -0,0 +1,9 @@
+  blockCleanerCommand:
+  - "/scripts/quick_reset.sh"
+  storageClass:
+    reclaimPolicy: Delete
```
I would leave out reclaim policy if you're going to use the default

cofyc commented:
Updated in https://github.com/kubernetes-incubator/external-storage/pull/789/commits/616d77d662e338dcc195d2a3368e1a4c30dff88b.

- Update docs.
- By default, don't set imagePullPolicy (will use default kubernetes imagePullPolicy).
- Remove `gce-cleanbyjobs.yaml`, add `baremetal-cleanbyjobs.yaml`.
- Comment other avaialbe block cleaner commands in `baremetal.yaml`.
- Leave out `reclaimPolicy: Delete` if we're going to use the default.

msau42 commented on local-volume/utils/update-helm-values/README.md:
```diff
@@ -0,0 +1,9 @@
+## update helm values
```
can you make this directory name be explicit about the before/after versions that this tool supports?

cofyc replied:
It supports its first version to local-volume-provisioner-v2.1.0, so I renamed it to `update-helm-values-pre-v2.2.0`. If you have a better name, let me change to it.

msau42 commented on local-volume/helm/examples/gce-retain.yaml:
```diff
@@ -0,0 +1,13 @@
+common:
+  # Beta PV.NodeAffinity field is used by default. If running against pre-1.10
+  # k8s version, the `useAlphaAPI` flag must be enabled in the configMap.
+  useAlphaAPI: false
```
Let's leave out any default values.

cofyc replied:
Done.

msau42 commented on local-volume/helm/provisioner/templates/provisioner-cluster-role-binding.yaml:
```diff
@@ -36,12 +36,12 @@ roleRef:
 kind: Role
 metadata:
-  name: jobsrole
+  name: local-storage-provisioner
```
A more descriptive name could be "local-storage-provisioner-jobs-role"

cofyc replied:
Done.

msau42 commented on local-volume/helm/provisioner/templates/provisioner.yaml:
```diff
@@ -1,39 +1,30 @@
-{{- range $tmp, $val := $classPath }}
-   {{- range $storageClass, $classConfig := $val }}
-{{ $storageClass }}:
+{{- range $classConfig := . -}}
```
Is it possible to put all this inline with the rest of the spec template instead of at the beginning?

cofyc replied:
Done, I removed variables aliases which I found not necessary too. It's hard to read diff of provisioner.yaml now, you can simply check [final version of provisioner](https://github.com/cofyc/external-storage/blob/refactor_heml_chart/local-volume/helm/provisioner/templates/provisioner.yaml).

I also added a helm chart test script here: https://github.com/cofyc/external-storage/tree/refactor_heml_chart/local-volume/helm/test

cofyc commented:
Updated in https://github.com/kubernetes-incubator/external-storage/pull/789/commits/1e4ab6437737ea7f9ea72cd4066c2b5a18a653c5:

- Rename update-helm-values to update-helm-values-pre-v2.2.0
- Add more tests in update-helm-values-pre-v2.2.0
- Leave out any default values in examples
- Rename role local-storage-provisioner to local-storage-provisioner-jobs-role
- Rename rolebinding local-storage-provisioner to local-storage-provisioner-jobs-rolebinding
- Inline all template definitions
- Remove all template varaibles alias
- Add helm tests and `make test-local-volume/helm`

cofyc commented:
I missed `daemonset.kubeConfigEnv` configuration, pushed a new commit https://github.com/kubernetes-incubator/external-storage/pull/789/commits/3958fb082bb8850d4928a95f4b76483b048b94a1.

msau42 commented on local-volume/helm/provisioner/templates/provisioner.yaml:
```diff
@@ -101,25 +63,39 @@ spec:
-
-{{- if $kubeConfigEnv }}
+            value: "{{ .Values.daemonset.image }}"
+{{- if .Values.daemonset.kubeConfigEnv }}
```
Can this be indented?

cofyc replied:
Of course.

msau42 reviewed (commented):
This lgtm! Please squash

Commit:
- Get rid of `engine` thing, it's better to use differnt values files for different environments
- Create job RBAC when necessary
- Able to create StorageClass object and configure it
- Add docs and examples for helm chart
- By default, don't set imagePullPolicy (will use default kubernetes imagePullPolicy)
- Add update-helm-values-pre-v2.2.0 for migrating helm chart values
- Inline all template definitions and remove all template varaibles aliases

Commit:
Add helm tests and make test-local-volume/helm

cofyc commented:
@msau42 Done. I squashed it into two commits:

- Improvements and refactorings: https://github.com/kubernetes-incubator/external-storage/pull/789/commits/68d5d1574c8cbb62eb8b02a4fbabe9033cb719ce
- Tests: https://github.com/kubernetes-incubator/external-storage/pull/789/commits/4ac87aad056ae496440d640c3705e63b9c7fb53e

msau42 commented:
/retest

cofyc commented:
seems like `/retest` does not trigger travis rebuild, I pushed again.

msau42 commented:
/lgtm
/assign @childsb
